### PR TITLE
Refactor `Assessment#questions.step` by using `next_unanswered`

### PR DESCRIPTION
**Pull request by allenwq**
@lowjoel  It's much cleaner now :-P

### CODE REVIEW

**Comment by lowjoel** [comment]
I think the documentation then needs to be refactored.

How does this change affect the number of SQL queries?

**Comment by allenwq** [comment]
@lowjoel I refactored the docs a bit, and the new way uses one less query.

File: app/models/concerns/course/assessment/questions_concern.rb [hunk 1]
```diff
@@ -30,24 +30,19 @@ def not_correctly_answered(submission)
...
-  # Return the question at the given step.
+  # Return the question at the given index. The next unanswered question will be returned if
+  # the question at the index is not accessible.
   #
   # @param [Course::Assessment::Submission] submission The submission which contains the answers.
   # @param [Fixnum] current_index The index of the question, it's zero based.
-  # @return [Array<Course::Assessment::Question>] The question at the given the step. The latest
-  #   unfinished question will be returned if the question at the step is not accessible.
+  # @return [Array<Course::Assessment::Question>] The question at the given index or next
+  #   unanswered question.
```
Comment (lowjoel) [inline app/models/concerns/course/assessment/questions_concern.rb]: `question, whichever comes first.`

**Comment by lowjoel** [comment]
Also, maybe we should remove all the not_X methods, have only the X methods, then do where.not(X) instead?

**Comment by allenwq** [comment]
@lowjoel Is there anyway to directly use where.not(X) ?  I think in that case for `not_correctly_answered` we will need to do `@assessment.questions.where.not(id: @assessment.questions.correctly_answered(@submisions))` ?

File: app/models/concerns/course/assessment/questions_concern.rb [hunk 2]
```diff
@@ -71,6 +66,10 @@ def next_unanswered(submission)
   # @param [Course::Assessment::Submission] submission The submission which contains the answers.
   # @return [Array<Course::Assessment::Question>] The questions which were correctly answered.
   def correctly_answered_questions(submission)
-    where(id: submission.answers.where(correct: true).select(:question_id))
+    where(id: correctly_answered_question_ids(submision))
+  end
+
+  def correctly_answered_question_ids(submission)
```
Comment (allenwq) [inline app/models/concerns/course/assessment/questions_concern.rb]: @lowjoel I have factored out the common parts of `not_correctly_answered` and `correctly_answered_questions`

**Comment by lowjoel** [comment]
@allenwq build failed

**Comment by allenwq** [comment]
@lowjoel it passed !

**Pull request merged**